### DISCUSSION TITLE
Support mapping composite/struct types

### DISCUSSION THREAD

**Discussion by beikov** [hibernate/hibernate-orm#5104] [Design Proposals]
Allow mapping an embeddable as composite/struct type. My suggestion is to require 1-2 steps to opt into this kind of mapping.

1. Add `@JdbcTypeCode( Types.STRUCT )` to the embedded attribute
2. Optionally define the name of the composite/struct type via `@Column(columnDefinition = "...")` - we derive a SQL type name from the simple embeddable class name

For an initial implementation, this should be enough, but I have some ideas what we could do to improve usability further

* Add annotation `@Struct(name = "...")` which can be annotated on the embeddable type. All type usages will use the composite/struct type by default
* Support DDL generation of composite/struct types: https://github.com/hibernate/hibernate-orm/discussions/3811
* Support controlling the physical order of columns/fields: https://github.com/hibernate/hibernate-orm/discussions/3989

My intention is to add this to the next feature release 6.2 as part of https://hibernate.atlassian.net/browse/HHH-15327.

**Comment by gavinking** [comment]
Looks good, but to me it feels like the main "usefulness" of structs is to be able to use them in `insert` lists and `update` lists, and to be able to do this consistently I think we would also need a way to "instantiate" a struct from within the language. Some sort of generalization of the `select new Foo()` thing that has existed forever.

**Reply by beikov** [reply]
Adding support for "instantiation" is probably even necessary to be able to work with these types in `insert` and `update`, so thanks that you mention it.

To me though, the main "usefulness" is, being able to map to this kind of DDL type at all. A composite/struct type has benefits on the DB side which don't immediately affect Hibernate users i.e. a non-Hibernate backend might want to use functions with composite/struct typed arguments for more type safety. When Hibernate enables mapping to these DDL types, users can finally switch to the more type safe approach. It's also nice that we might acquire new users that couldn't or didn't want to use Hibernate so far since it was lacking support for composite/struct types.

**Reply by gavinking** [reply]
Sure, I don't disagree with that at all.

**Comment by gavinking** [comment]
When I look at our code base today, this is starting to feel like a missing feature. We should do it. I think it’s a lot more natural than SQL arrays, which we do have.

**Reply by beikov** [reply]
FYI https://hibernate.atlassian.net/browse/HHH-15327
